File: 02/java/LeetCode.java
import java.util.TreeSet;
import java.util.TreeMap;
import java.util.ArrayList;
import java.util.List;
import java.util.LinkedList;

public class LeetCode {
	public LeetCode() {}

	private boolean num20(String s) {
		ArrayStack<Character> stack = new ArrayStack<>();

	    int len = s.length();
	    for( int i = 0; i < len; i++ ) {
	    	char c = s.charAt(i);
	    	if( c == '{' || c == '[' || c =='(' ) {
	    		stack.push( c );
	    	} else {
	    		if( stack.isEmpty() ) return false;
	    		char top = stack.pop();
	    		if( c == '}' && top != '{' ) return false;
	    		if( c == ']' && top != '[' ) return false;
	    		if( c == ')' && top != '(' ) return false;
	    	}
	    }

	    return stack.isEmpty();
	}
	public void test20(String s) {
		Boolean res = num20(s);
		System.out.println( res );
	}

	private class ListNode {
		int val;
		ListNode next;
		ListNode(int x) { val = x; }
		@Override
		public String toString() {
			StringBuilder s = new StringBuilder();
			ListNode cur = this;
			while( cur != null ) {
				s.append(cur.val + " => ");
				cur = cur.next;
			}
			s.append("NULL");
			return s.toString();
		}
	}
	private ListNode num203(ListNode head, int val) {
		/*while( head != null && head.val == val ) {
			ListNode delNode = head;
			head = head.next;
			delNode.next = null;
		}
		if( head == null ) return head;

		ListNode prev = head;
		while( prev.next != null ) {
			if( prev.next.val == val ) {
				ListNode delNode = prev.next;
				prev.next = delNode.next;
				delNode.next = null;
			} else {
				prev = prev.next;
			}
		}

		return head;*/

		/*ListNode dummyHead = new ListNode(-1);
		dummyHead.next = head;

		ListNode prev = dummyHead;
		while( prev.next != null ) {
			if( prev.next.val == val ) {
				ListNode delNode = prev.next;
				prev.next = delNode.next;
				delNode.next = null;
			} else {
				prev = prev.next;
			}
		}

		return dummyHead.next;*/

		if( head == null ) return null;
		head.next = num203(head.next, val);
		return head.val == val ? head.next : head;
	}
	public void test203(int[] nums, int val) {
		ListNode dummyHead = new ListNode(0);
		ListNode cur = dummyHead;

		for( int i=0; i<nums.length; i++ ) {
			cur.next = new ListNode(nums[i]);
			cur = cur.next;
		}

		ListNode head = dummyHead.next;
		System.out.println("入  " + head + "  删除" + val);
		head = num203(head, val);
		System.out.println("出  " + head);
	}

	public int num804(String[] words) {
		TreeSet<String> set = new TreeSet<>();
		String[] codes = {".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};

		for(String word : words) {
			StringBuilder res = new StringBuilder();
			for(int i=0; i<word.length(); i++) {
				res.append(codes[word.charAt(i) - 'a']);
			}
			set.add(res.toString());
		}

		return set.size();
	}

	public int[] num349(int[] nums1, int[] nums2) {
		TreeSet<Integer> set = new TreeSet<>();
		for( int num : nums1 ) set.add(num);

		ArrayList<Integer> list = new ArrayList<>();
		for( int num : nums2 ) {
			if( set.contains(num) ) {
				list.add(num);
				set.remove(num);
			}
		}

		int size = list.size();
		int[] res = new int[size];
		for( int i=0; i<size; i++ ) res[i] = list.get(i);

		return res;
	}
	
	public int[] num350(int[] nums1, int[] nums2) {
		TreeMap<Integer, Integer> map = new TreeMap<>();
		for( int num : nums1 ) {
			int val = map.containsKey(num) ? map.get(num)+1 : 1;
			map.put(num, val);
		}

		ArrayList<Integer> list = new ArrayList<>();
		for( int num : nums2 ) {
			if( map.containsKey(num) ) {
				list.add(num);
				map.put(num, map.get(num) - 1);
				if( map.get(num) == 0 ) map.remove(num);
			}
		}

		int size = list.size();
		int[] res = new int[size];
		for( int i=0; i<size; i++ ) res[i] = list.get(i);

		return res;
    }

	private class Freq implements Comparable<Freq> {
		public int e, freq;

		public Freq(int e, int freq) {
			this.e = e;
			this.freq = freq;
		}

		@Override
		public int compareTo(Freq another) {
			if( this.freq < another.freq ) {
				return 1;
			} else if ( this.freq > another.freq ) {
				return -1;
			} else {
				return 0;
			}
		}
	}
	public int[] num347(int[] nums, int k) {
		TreeMap<Integer, Integer> map = new TreeMap<>();
		for( int num: nums ) {
			if( map.containsKey(num) ) {
				map.put(num, map.get(num) + 1);
			} else {
				map.put(num, 1);
			}
		}

		PriorityQueue<Freq> pq = new PriorityQueue<>();
		for( int key: map.keySet() ) {
			if( pq.getSize() < k ) {
				pq.enqueue(new Freq(key, map.get(key)));
			} else if( map.get(key) > pq.getFront().freq ) {
				pq.dequeue();
				pq.enqueue(new Freq(key, map.get(key)));
			}
		}

		// LinkedList<Integer> res = new LinkedList<>();
		int len = pq.getSize();
		int[] res = new int[len];
		for (int i = len-1; i >= 0; i--) {
			res[i] = pq.dequeue().e;
		}

		return res;
	}
}

File: 02/js/leetcode.js
const BST = require('./BST');

const num20 = s => {
	var stack = [];
    var len = s.length;
    var maps = { '[': ']', '{': '}', '(': ')' };
    for( var i = 0; i < len; i++ ) {
        var c = s.charAt(i);
        if( c === '{' || c === '[' || c === '(' ) {
            stack.push( c );
        } else {
            if( !stack.length ) return false;
            var top = stack.pop();
            if( maps[top] !== c ) return false;
        }
    }
    
    if( !stack.length ) return true;
    return false;
};
exports.test20 = s => console.log( num20(s) );

function ListNode(val) {
    this.val = val;
    this.next = null;
    this.toString = () => {
        let res = '';
        let cur = this;

        while( cur ) {
            res += `${cur.val} => `;
            cur = cur.next;
        }
        return `${res}NULL`;
    };
}
const num203 = (head, val) => {
    /*while( head && head.val === val ) {
        const delNode = head;
        head = delNode.next;
        delNode.next = null;
    }
    if( !head ) return head;

    let prev = head;
    while( prev.next ) {
        if( prev.next.val === val ) {
            const delNode = prev.next;
            prev.next = delNode.next;
            delNode.next = null;
        } else {
            prev = prev.next;
        }
    }
    return head;*/

    /*const dummyHead = new ListNode(0);
    dummyHead.next = head;

    let prev = dummyHead;
    while( prev.next ) {
        if( prev.next.val === val ) {
            const delNode = prev.next;
            prev.next = delNode.next;
            delNode.next = null;
        } else {
            prev = prev.next;
        }
    }
    return dummyHead.next;*/

    if( !head ) return null;
	head.next = num203(head.next, val);
	return head.val == val ? head.next : head;
};
exports.test203 = (nums, val) => {
    const dummyHead = new ListNode(-1);
    let cur = dummyHead;
    nums.forEach(item => {
        cur.next = new ListNode(item);
        cur = cur.next;
    });

    let head = dummyHead.next;
    console.log("入  " + (head ? head.toString() : head) + "  删除" + val);
    head = num203(head, val);
    console.log("出  " + (head ? head.toString() : head));
};
export const num804 = words => {
    const set = new Set();
    const baseNum = 'a'.charCodeAt();
    const codes = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."];

    for(let word of words) {
        const res = [];
        word.split('').forEach(letter => res.push(codes[letter.charCodeAt()-baseNum]));
        set.add(res.join(''));
    }
    return set.size;
};

exports.test938 = (root, L, R) => {
    var rangeSumBST = function(root, L, R) {
        if( !root ) return 0;

        if( root.val < L ) return rangeSumBST(root.right, L, R);
        if( root.val > R ) return rangeSumBST(root.left, L, R);
        if( root.val >= L && root.val <= R ) {
            return root.val + rangeSumBST(root.left, L, R) + rangeSumBST(root.right, L, R);
        }
    };
};
// 合并二叉树
exports.test617 = () => {
    var mergeTrees = function(t1, t2) {
        if( !t1 && !t2 ) return null;
        if( !t1 ) return t2;
        if( !t2 ) return t1;

        const node = new TreeNode(t1.val + t2.val);
        node.left = mergeTrees(t1.left, t2.left);
        node.right = mergeTrees(t1.right, t2.right);

        return node;
    };
};
// 翻转二叉树
exports.test226 = () => {
    var invertTree = function(root) {
        if( !root ) return null;

        const node = new TreeNode(root.val);
        const lNode = invertTree(root.left);
        node.right = lNode;
        const rNode = invertTree(root.right);
        node.left = rNode;

        return node;
    };
};
// 二叉树最大深度
exports.test104 = () => {
    var maxDepth = function(root) {
        if( !root ) return 0;

        /*let depth = 0;
        const arr = [];

        root.val = 1;
        arr.push(root);

        while( arr.length ) {
            const node = arr.shift();
            depth = node.val;

            if( node.left ) {
                node.left.val = depth + 1;
                arr.push(node.left);
            }
            if( node.right ) {
                node.right.val = depth + 1;
                arr.push(node.right);
            }
        }

        return depth;*/

        const lDepth = maxDepth(root.left);
        const rDepth = maxDepth(root.right);
        return Math.max(lDepth, rDepth) + 1;
    };
};
// 二叉树中序遍历
exports.test94 = () => {
    var inorderTraversal = function(root) {
        /*let numbers = [];
        if( !root ) return numbers;

        numbers = inorderTraversal(root.left);
        numbers.push(root.val);
        numbers = numbers.concat(inorderTraversal(root.right));

        return numbers;*/

        if( !root ) return [];

        const arr = [];
        const numbers = [];
        arr.push( root );

        let cur = root;
        while( arr.length ) {
            if( !cur ) {
                const node = arr.pop();
                numbers.push( node.val );

                cur = node.right;
                if( cur ) arr.push(cur);
            } else {
                cur = cur.left;
                if( cur ) arr.push(cur);
            }
        }

        return numbers;
    };
};
// 二叉树层次遍历
exports.test107 = () => {
    var levelOrderBottom = function(root) {
        if( !root ) return [];

        root.depth = 0;
        const arr = [root];
        const numbers = [[]];

        while( arr.length ) {
            const node = arr.shift();
            numbers[node.depth].push(node.val);

            const index = node.depth + 1;
            if( !numbers[index] && (node.left || node.right) ) numbers[index] = [];

            if( node.left ) {
                node.left.depth = index;
                arr.push(node.left);
            }
            if( node.right ) {
                node.right.depth = index;
                arr.push(node.right);
            }
        }
        return numbers.reverse();
    };
};
